### PR TITLE
Enable open/close/open flow for device/module clients

### PR DESCRIPTION
**Pull request by timtay-microsoft**
This functionality already almost worked. It just requires the retry delegating handler not disposing itself on close and updating the _isOpen flag on close.

Add e2e tests that confirm you can still perform basic twin/telemetry/method operations after re-opening a client.

### CODE REVIEW

File: iothub/device/src/Pipeline/RetryDelegatingHandler.cs [hunk 1]
```diff
@@ -454,6 +454,8 @@ public override async Task CloseAsync(CancellationToken cancellationToken)
             if (Logging.IsEnabled)
                 Logging.Enter(this, cancellationToken, nameof(CloseAsync));
 
+            _isOpen = false;
```
Comment (abhipsaMisra) [inline iothub/device/src/Pipeline/RetryDelegatingHandler.cs:457]: should this flag be set after it has been disposed?

Comment (timtay-microsoft) [reply iothub/device/src/Pipeline/RetryDelegatingHandler.cs:457]: It already sets this flag to false, yeah.

**Review by abhipsaMisra — APPROVED** [review]

**Review by tmahmood-microsoft — APPROVED** [review]

File: SDK v2 migration guide.md [hunk 2]
```diff
@@ -61,6 +61,7 @@ to migrate to version 2.x when they have the chance. For more details on LTS rel
 - The device and module clients now have a property (e.g., `IotHubDeviceClient.ConnectionStatusInfo`)with the latest connection status information on it, eliminating the need for a connection status callback method to cache the latest values.
 - Remote certificate validation is no natively longer supported for AMQP web socket connections. Supprted workaround is to provide a client web socket instance in the client options.
 - Added support for setting a client web socket instance in the client options so that users can have better control over AMQP web socket connections.
+- The client can now be re-opened after it has been closed. It cannot be re-opened after it has been disposed, though. Also, subscriptions do not carry over when the client is re-opened.
```
Comment (drwill-ms) [inline SDK v2 migration guide.md:64]: What would it take for subscriptions to carry over? Do they in Java?

**Review by drwill-ms — APPROVED** [review]

File: SDK v2 migration guide.md (hunk 2 above)
Comment (timtay-microsoft) [reply SDK v2 migration guide.md:64]: I don't have any strong feelings about carrying them over or not on re-open. In Java, they only persist across reconnection, not open/close/open.

File: iothub/device/src/IotHubBaseClient.cs [hunk 3]
```diff
@@ -116,6 +116,11 @@ public void SetRetryPolicy(IRetryPolicy retryPolicy)
         /// <summary>
         /// Open the client instance. Must be done before any operation can begin.
         /// </summary>
+        /// <remarks>
+        /// This client can be re-opened after it has been closed, but cannot be re-opened after it has
```
Comment (patilsnr) [inline iothub/device/src/IotHubBaseClient.cs:120]: Super (and perhaps subjective) nit, but `are not persisted` reads a bit awkward to me -- `do not persist` may be better.

**Review by patilsnr — APPROVED** [review]

**Comment by timtay-microsoft** [comment]
Update the reconnection sample before merging

**Pull request merged**